runtime/permissions.rs

```diff
@@ -115,6 +115,62 @@ impl Default for PermissionState {
   }
 }
 
+/// A normalized environment variable name. On Windows this will
+/// be uppercase and on other platforms it will stay as-is.
+#[derive(Clone, Eq, PartialEq, Hash, Debug, Default)]
+struct EnvVarName {
+  inner: String,
+}
+
+impl EnvVarName {
+  pub fn new(env: impl AsRef<str>) -> Self {
+    Self {
+      inner: if cfg!(windows) {
+        env.as_ref().to_uppercase()
+      } else {
+        env.as_ref().to_string()
+      },
+    }
+  }
+}
+
+impl AsRef<str> for EnvVarName {
+  fn as_ref(&self) -> &str {
+    self.inner.as_str()
+  }
+}
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug, Default)]
+pub struct PermissionPath {
+  original_path: PathBuf,
+  resolved_path: PathBuf,
+}
+
+impl PermissionPath {
+  pub fn new(path: impl AsRef<Path>) -> Self {
+    Self {
+      original_path: path.as_ref().to_path_buf(),
+      resolved_path: resolve_from_cwd(path.as_ref()).unwrap(),
+    }
+  }
+
+  pub fn starts_with(&self, other_path: &PermissionPath) -> bool {
+    self.resolved_path.starts_with(&other_path.resolved_path)
+  }
+
+  pub fn as_path(&self) -> &Path {
+    // Ensure the original path is returned in order to prevent providing
+    // the resolved path to another method in the permissions and exposing
+    // the CWD
+    self.original_path.as_path()
+  }
```

Comment:
I didn't add a test for this because it was painful, but I just maybe thought of a better way of doing it, so I will try again right now.